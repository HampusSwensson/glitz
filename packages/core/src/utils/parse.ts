import { Declarations } from '@glitz/type';

export function parseDeclarationBlock(declarations: Declarations) {
  let block = '';
  let property: string;
  for (property in declarations) {
    const value = declarations[property];
    if (typeof value === 'object') {
      for (const fallback of value) {
        if (block) {
          block += ';';
        }
        block += parseDeclaration(property, fallback);
      }
    } else {
      if (block) {
        block += ';';
      }
      block += parseDeclaration(property, value);
    }
  }
  return block;
}

export function parseDeclaration(property: string, value?: string | number) {
  return `${hyphenateProperty(property)}:${value}`;
}

// Accept both camel cased and capitalized vendor properties
const hyphenateRegex = /(?:^(ms|moz|webkit))|[A-Z]/g;
const propertyCache: { [property: string]: string } = {};

export function hyphenateProperty(property: string) {
  return property in propertyCache
    ? propertyCache[property]
    : (propertyCache[property] = property.replace(hyphenateRegex, '-$&').toLowerCase());
}
