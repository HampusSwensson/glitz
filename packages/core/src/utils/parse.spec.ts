import { parseDeclaration, parseDeclarationBlock } from './parse';

describe('parse', () => {
  it('parses declaration', () => {
    expect(parseDeclaration('color', 'red')).toBe('color:red');
    expect(parseDeclaration('backgroundColor', 'green')).toMatchSnapshot();

    expect(parseDeclaration('msOverflowStyle', 'auto')).toMatchSnapshot();
    expect(parseDeclaration('MozAppearance', 'button')).toMatchSnapshot();
    expect(parseDeclaration('WebkitBorderBefore', 'red')).toMatchSnapshot();
  });
  it('parses block', () => {
    expect(parseDeclarationBlock({ color: 'red' })).toMatchSnapshot();
    expect(parseDeclarationBlock({ backgroundColor: 'green' })).toMatchSnapshot();
    expect(parseDeclarationBlock({ color: 'blue', backgroundColor: 'transparent' })).toMatchSnapshot();
  });
  it('parses fallback values', () => {
    expect(parseDeclarationBlock({ color: ['red', 'green'] })).toMatchSnapshot();
  });
});
