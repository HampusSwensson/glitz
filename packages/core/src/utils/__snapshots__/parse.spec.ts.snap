// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse parses block 1`] = `"color:red"`;

exports[`parse parses block 2`] = `"background-color:green"`;

exports[`parse parses block 3`] = `"color:blue;background-color:transparent"`;

exports[`parse parses declaration 1`] = `"background-color:green"`;

exports[`parse parses declaration 2`] = `"-ms-overflow-style:auto"`;

exports[`parse parses declaration 3`] = `"-moz-appearance:button"`;

exports[`parse parses declaration 4`] = `"-webkit-border-before:red"`;

exports[`parse parses fallback values 1`] = `"color:red;color:green"`;
